internal/services/mobilenetwork/mobile_network_attached_data_network_resource.go

```diff
@@ -0,0 +1,549 @@
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
+)
+
+type AttachedDataNetworkModel struct {
+	MobileNetworkDataNetworkName         string                   `tfschema:"mobile_network_data_network_name"`
+	MobileNetworkPacketCoreDataPlaneId   string                   `tfschema:"mobile_network_packet_core_data_plane_id"`
+	DnsAddresses                         []string                 `tfschema:"dns_addresses"`
+	Location                             string                   `tfschema:"location"`
+	NaptConfiguration                    []NaptConfigurationModel `tfschema:"network_address_port_translation"`
+	Tags                                 map[string]interface{}   `tfschema:"tags"`
+	UserEquipmentAddressPoolPrefix       []string                 `tfschema:"user_equipment_address_pool_prefixes"`
+	UserEquipmentStaticAddressPoolPrefix []string                 `tfschema:"user_equipment_static_address_pool_prefixes"`
+	UserPlaneAccessIPv4Address           string                   `tfschema:"user_plane_access_ipv4_address"`
+	UserPlaneAccessIPv4Gateway           string                   `tfschema:"user_plane_access_ipv4_gateway"`
+	UserPlaneAccessIPv4Subnet            string                   `tfschema:"user_plane_access_ipv4_subnet"`
+	UserPlaneAccessName                  string                   `tfschema:"user_plane_access_name"`
+}
+
+type NaptConfigurationModel struct {
+	PinholeLimits           int64            `tfschema:"pinhole_maximum_number"`
+	IcmpPinholeTimeout      int64            `tfschema:"icmp_pinhole_timeout_in_seconds"`
+	TcpPinholeTimeout       int64            `tfschema:"tcp_pinhole_timeout_in_seconds"`
+	UdpPinholeTimeout       int64            `tfschema:"udp_pinhole_timeout_in_seconds"`
+	PortRange               []PortRangeModel `tfschema:"port_range"`
+	TcpReuseMinimumHoldTime int64            `tfschema:"tcp_port_reuse_minimum_hold_time_in_seconds"`
+	UdpReuseMinimumHoldTime int64            `tfschema:"udp_port_reuse_minimum_hold_time_in_seconds"`
+}
+
+type PortRangeModel struct {
+	MaxPort int64 `tfschema:"max_port"`
+	MinPort int64 `tfschema:"min_port"`
+}
+
+type AttachedDataNetworkResource struct{}
+
+var _ sdk.ResourceWithUpdate = AttachedDataNetworkResource{}
+
+func (r AttachedDataNetworkResource) ResourceType() string {
+	return "azurerm_mobile_network_attached_data_network"
+}
+
+func (r AttachedDataNetworkResource) ModelObject() interface{} {
+	return &AttachedDataNetworkModel{}
+}
+
+func (r AttachedDataNetworkResource) IDValidationFunc() pluginsdk.SchemaValidateFunc {
+	return attacheddatanetwork.ValidateAttachedDataNetworkID
+}
+
+func (r AttachedDataNetworkResource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"mobile_network_data_network_name": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
```

Comment:
As above, this should be the ID of the related resource:
```suggestion
		"mobile_network_data_network_id": {
			Type:         pluginsdk.TypeString,
			Required:     true,
			ForceNew:     true,
			ValidateFunc: datanetwork.ValidateDataNetworkID,
		},
```
From this we can extract the correct name